src/proof/alethe/alethe_node_converter.cpp

```diff
@@ -21,39 +21,61 @@
 namespace cvc5::internal {
 namespace proof {
 
-Node AletheNodeConverter::preConvert(Node n)
-{
-  Kind k = n.getKind();
-  if (k == kind::SKOLEM || k == kind::BOOLEAN_TERM_VARIABLE)
-  {
-    Trace("alethe-conv") << "AletheNodeConverter: [PRE] handling skolem " << n
-                         << "\n";
-    Node wi = SkolemManager::getWitnessForm(n);
-    // true skolem with witness form, just convert that
-    if (!wi.isNull())
-    {
-      Trace("alethe-conv") << "AletheNodeConverter: ..skolem " << n
-                           << " has witness form " << wi << "\n";
-      return wi;
-    }
-    // purification skolem, so we simply retrieve its original form and convert
-    // that
-    Node oi = SkolemManager::getOriginalForm(n);
-    AlwaysAssert(!oi.isNull());
-    Trace("alethe-conv")
-        << "AletheNodeConverter: ..pre-convert to original form " << oi << "\n";
-    return oi;
-  }
-  return n;
-}
-
 Node AletheNodeConverter::postConvert(Node n)
 {
   NodeManager* nm = NodeManager::currentNM();
   Kind k = n.getKind();
-  AlwaysAssert(k != kind::SKOLEM && k != kind::BOOLEAN_TERM_VARIABLE);
   switch (k)
   {
+    case kind::SKOLEM:
+    case kind::BOOLEAN_TERM_VARIABLE:
+    {
+      Trace("alethe-conv") << "AletheNodeConverter: handling skolem " << n
+                           << "\n";
+      Node wi = SkolemManager::getOriginalForm(n);
+      Trace("alethe-conv") << "AletheNodeConverter: ..original: " << wi << "\n";
+      if (wi == n)
+      {
+        // if it is not a purification skolem, maybe it has a witness skolem
+        wi = SkolemManager::getWitnessForm(n);
+      }
+      // skolem with witness form, just convert that
+      if (!wi.isNull() && wi != n)
```

Comment:
if you use "else if" you can loose the wi != n condition

Reply:
Hum, the call in line 40 can modify `wi`, so using an `else if` and keep the same behavior will depend on what `getWitnessForm` is doing, i.e., if whenever it returns a non-null term it is not the same as `n`. This should always be the case, right @ajreynol ?

Reply:
Yes, this is correct.

Although note that `getWitnessForm` may be eliminated in the near future.

Reply:
OK, thanks! And about eliminating `getWitnessForm`, well, as long as there is an equivalent functionality...

Reply:
It will be another skolem function identifier `QUANTIFIERS_SKOLEMIZE`.

Reply:
See https://github.com/cvc5/cvc5/pull/9536.

Reply:
Given the big change that #9536 introduces I'd like to merge this in first, as it fixes the aforementioned issues. And also provides a use case for skolems that #9536 must not break.

Reply:
Sure, then you can formalize the skolem as a choice term here in whatever way you want.

Reply:
I have removed the thing I was playing around for array extensionality. The handling for quantifier skolemization however must be that one (or something equivalent to it) in Alethe's current form.